sw/device/lib/testing/flash_ctrl_testutils.c

```diff
@@ -0,0 +1,182 @@
+ * Returns the result of transaction_end.
+ */
+int flash_ctrl_testutils_write_page(
+    dif_flash_ctrl_state_t *flash_state, uint32_t byte_address,
+    const uint32_t *data, dif_flash_ctrl_partition_type_t partition_type,
+    uint32_t word_count) {
+  dif_flash_ctrl_transaction_t transaction = {.byte_address = byte_address,
+                                              .op = kDifFlashCtrlOpProgram,
+                                              .partition_type = partition_type,
+                                              .partition_id = 0x0,
+                                              .word_count = 0x0};
+  uint32_t words_written = 0;
+  int retval = 0;
+  while (words_written < word_count) {
+    uint32_t words_to_write = ((word_count - words_written) < WORDS_PER_ACCESS)
+                                  ? (word_count - words_written)
+                                  : WORDS_PER_ACCESS;
+    transaction.word_count = words_to_write;
+    transaction.byte_address =
+        byte_address + (sizeof(uint32_t) * words_written);
+    CHECK_DIF_OK(dif_flash_ctrl_start(flash_state, transaction));
+    CHECK_DIF_OK(dif_flash_ctrl_prog_fifo_push(flash_state, words_to_write,
+                                               data + words_written));
+    retval |= flash_ctrl_testutils_wait_transaction_end(flash_state);
+    words_written += words_to_write;
+  }
+
+  return retval;
+}
+
+/**
+ * Erases and Programs the page at byte_address.
+ * Calls the erase and write functions in turn.
+ * Returns the combined result of the transaction_ends.
+ */
+int flash_ctrl_testutils_erase_and_write_page(
+    dif_flash_ctrl_state_t *flash_state, uint32_t byte_address,
+    const uint32_t *data, dif_flash_ctrl_partition_type_t partition_type,
+    uint32_t word_count) {
+  int retval = flash_ctrl_testutils_erase_page(flash_state, byte_address,
+                                               partition_type);
+  retval |= flash_ctrl_testutils_write_page(flash_state, byte_address, data,
+                                            partition_type, word_count);
+  return retval;
+}
+
+/**
+ * Reads the page at byte_address.
+ * Returns the result of transaction_end.
+ */
+int flash_ctrl_testutils_read_page(
+    dif_flash_ctrl_state_t *flash_state, uint32_t byte_address,
+    uint32_t *data_out, dif_flash_ctrl_partition_type_t partition_type,
+    uint32_t word_count, uint32_t delay) {
+  dif_flash_ctrl_transaction_t transaction = {.byte_address = byte_address,
+                                              .op = kDifFlashCtrlOpRead,
+                                              .partition_type = partition_type,
+                                              .partition_id = 0x0,
+                                              .word_count = word_count};
+
+  // Read Page.
+  CHECK_DIF_OK(dif_flash_ctrl_start(flash_state, transaction));
+  // Optional delay to allow for read fifo fill testing.
+  if (delay > 0)
+    usleep(delay);
```

Comment:
The `delay` is unsigned so I would remove this `if`.
You need to rebase your branch, because the `usleep` has been replaced by `busy_spin_micros`